clients/src/main/resources/common/message/ProduceResponse.json

```diff
@@ -48,10 +48,15 @@
           "about": "The timestamp returned by broker after appending the messages. If CreateTime is used for the topic, the timestamp will be -1.  If LogAppendTime is used for the topic, the timestamp will be the broker local time when the messages are appended." },
         { "name": "LogStartOffset", "type": "int64", "versions": "5+", "default": "-1", "ignorable": true,
           "about": "The log start offset." },
-        { "name":  "ErrorRecords", "type":  "[]int32", "versions": "8+", "default": "", "ignorable":  true,
-          "about":  "The relative offset of records that cause the batch to be dropped"},
+        { "name": "ErrorRecords", "type": "[]RelativeOffsetAndErrorMessage", "versions":  "8+", "default": "", "ignorable": true,
+          "about": "The relative offsets of records that caused the batch to be dropped", "fields": [
+          { "name":  "RelativeOffset", "type": "int32", "versions":  "8+",
+            "about":  "The relative offset of the record that cause the batch to be dropped" },
+          { "name": "RelativeOffsetErrorMessage", "type": "string", "versions":  "8+",
```

Comment:
Do we want this to be nullable? Same for ErrorMessage below. We'll need to make a similar change to the old style spec.

Reply:
The error message is great, but it can't be used programmatically on the client/app, so we'll also need the 16-bit error code here.

Reply:
@edenhill I've thought about this when proposing KIP-467 and the reason I did not try to encode the 16 bit error code is that:

1. At the moment, from all the possible errors that can be returned from the broker, except the InvalidRecordException all others would cause the response to be returned immediately, which means that we would not have more than one error code that would trigger the response to be rejected.

2. For the only possible InvalidRecordException that may be co-exist with other exceptions in the rejected response, that would cause the error records to be trimmed, whereas all other error codes would still cause the whole batch to be dropped. Therefore we've decided to make the broker-side logic to execute the following: when any non-InvalidRecordException is thrown, since that would cause the whole batch to be dropped we would just return that error code on the per-partition and ignore any other InvalidRecordException; if the only exception thrown is InvalidRecordException on specific record(s), then we encode it on per-partition level.

That being said, if people feel that in the future or even now we would like to change the broker-side logic and it's worth paying the 16bits in addition to 8bit for encoding an empty string, then I'm fine doing that in this PR as well.

@hachikuji @tuvtran @junrao please let me know.

Reply:
Thanks for explaining your reasoning @guozhangwang.

I think the key here is "At the moment": having an error code field in this response would allow any future error code to be added without having to bump the protocol version, more specifically the range of returned error codes for the per-message errors should be left undefined as this is a means for communication between some future message verification mechanism on the broker and the application producing the data, rather than between the broker and the producer.

For space efficiency I'd suggest using a varint, but I doubt it matters in practice since these errors should be rare, and the two bytes of error code are just a fraction of the error message. And the erroneous Produce response is most likely far smaller than the request, or a fetch response for that matter.

Reply:
@edenhill Honestly I'm still not convinced of the future usage of per-record error codes -- atm we have not observed concrete error codes that would split within a single batch that would like to be handled differently by clients. If we do find out that become a common scenario to support we can still always bump up the version, which is not a very costly evolution anyways. On the other hand, since we do not have varint in place yet we would still need to encode 2bytes with the empty string: note that the per-record error message would be commonly empty in practice also.

Another thing worth mentioning is that even if we add the fields now, any new error code beyond InvalidRecord that we'd like to add in the future would still require the clients to upgrade because otherwise they cannot recognize them anyways.

Reply:
@guozhangwang Okay, sounds good, I'll rest my case.